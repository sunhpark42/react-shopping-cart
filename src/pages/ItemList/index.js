import React from 'react';
import { useHistory } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { addItemToCart } from '../../store/cartReducer';
import { API } from '../../utils';
import { MESSAGE, PATH } from '../../constants';
import { Grid, Card, IconButton } from '../../components/shared';
import { Description, Price } from './style';
import { ReactComponent as CartIcon } from '../../assets/icons/cart.svg';

const ItemList = () => {
  const list = useSelector(state => state.reducer.itemList);
  const dispatch = useDispatch();
  const history = useHistory();

  const goDetailPage = id => () => {
    history.push(`${PATH.GOODS_DETAIL}?id=${id}`);
  };

  const addCart = id => async () => {
    try {
      const newCartItem = await API.addItemToCart(id);

      dispatch(addItemToCart(newCartItem));
      alert(MESSAGE.SUCCESS_ADD_ITEM_TO_CART);
    } catch (error) {
      console.error(error.message);
      alert(MESSAGE.FAIL_ADD_ITEM_TO_CART);
    }
  };

  return (
    <Grid col="4">
      {Object.values(list).map(({ product_id: id, name, image_url, price }) => {
        return (
          <Card
            key={id}
            title={name}
            thumbnail={{ image: image_url, alt: name }}
            onClick={goDetailPage(id)}
            description={
              <Description>
                <Price>{price.toLocaleString('ko-KR')} 원</Price>
                <IconButton
                  size="medium"
                  ariaLabel={`${name}을 장바구니에 담기`}
                  onClick={addCart(id)}
                >
                  <CartIcon />
                </IconButton>
              </Description>
            }
          />
        );
      })}
    </Grid>
  );
};

export default ItemList;
