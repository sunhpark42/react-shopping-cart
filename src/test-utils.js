// test-utils.js
import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { combineReducers, createStore } from 'redux';
import { Provider } from 'react-redux';
// Import your own reducer

import { reducer } from './store';
import { cartReducer } from './store/cartReducer';

const rootReducer = combineReducers({ cartReducer, reducer });

function render(
  ui,
  { initialState, store = createStore(rootReducer, initialState), ...renderOptions } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
