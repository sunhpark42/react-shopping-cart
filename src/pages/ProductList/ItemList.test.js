import React from 'react';
import ItemList from '.';
import { render, fireEvent, waitFor } from '../../test-utils';
import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';

const initialState = {
  cartReducer: {
    cart: [
      {
        id: 1,
        name: '[든든] 유부 슬라이스 500g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11153-main-01.png',
        price: 4900,
        quantity: 4,
        checked: true,
      },
    ],
  },
  reducer: {
    itemList: [
      {
        id: 1,
        name: '[든든] 유부 슬라이스 500g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11153-main-01.png',
        price: 4900,
      },
      {
        id: 2,
        name: '[든든] 진맛살 1kg',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11157-main-01.png',
        price: 7100,
      },
      {
        id: 3,
        name: '[든든] 신-키리모찌 1kg',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11161-main-01.png',
        price: 12000,
      },
      {
        id: 4,
        name: '[든든] 흑곤약 250g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11263-main-01.png',
        price: 1300,
      },
      {
        id: 5,
        name: '[든든] 흑곤약 1.2kg',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11277-main-01.png',
        price: 4000,
      },
      {
        id: 6,
        name: '[든든] 곤약말이 시라타끼 200g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11280-main-01.png',
        price: 1100,
      },
      {
        id: 7,
        name: '[든든] 기분 시로가마보코 160g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11315-main-01.png',
        price: 2800,
      },
      {
        id: 8,
        name: '[든든] 냉동조미유부삼각 (60장) 1kg',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11323-main-01.png',
        price: 8400,
      },
      {
        id: 9,
        name: '[든든] 시노다마끼 유부야채말이 (15g*40입) 600g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11328-main-01.png',
        price: 9000,
      },
      {
        id: 10,
        name: '[든든] 기분 오뎅세트 433g',
        image: 'https://cdn-mart.baemin.com/goods/custom/20200525/11331-main-01.png',
        price: 8900,
      },
    ],
    orderList: [],
  },
};

describe('상품목록 페이지', () => {
  it('store에 있는 products가 제대로 렌더링 되는지 확인한다.', () => {
    const utils = render(<ItemList />, { initialState });

    expect(utils.container).toMatchSnapshot();
  });

  it('장바구니 버튼 클릭시, 장바구니에 상품이 추가되었습니다. 가 뜬다.', async () => {
    const utils = render(<ItemList />, { initialState });
    const cartButton = utils.getByRole('button', {
      name: /\[든든\] 진맛살 1kg을 장바구니에 담기/i,
    });

    fireEvent.click(cartButton);

    const alertSpy = jest.spyOn(window, 'alert');
    alertSpy.mockImplementation(jest.fn(() => true));
    await waitFor(() => expect(alertSpy).toBeCalled());
  });

  it('아이템 이미지 클릭 시, "/goods/detail/?id=${id}"로 이동한다.', () => {
    const utils = render(
      <Router>
        <ItemList />
      </Router>,
      { initialState },
    );
    const itemImage = utils.getByRole('img', {
      name: /\[든든\] 유부 슬라이스 500g/i,
    });

    fireEvent.click(itemImage);
    expect(window.location.href).toBe('http://localhost/goods/detail/?id=1');
  });
});
