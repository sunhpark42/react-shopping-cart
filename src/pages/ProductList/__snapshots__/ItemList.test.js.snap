// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`상품목록 페이지 store에 있는 products가 제대로 렌더링 되는지 확인한다. 1`] = `
<div>
  <section
    class="sc-bdvvaa epvVwG"
  >
    <div
      class="sc-eCImvq kFPXzk"
    >
      <div
        class="sc-gsDJrp"
      >
        <div
          class="sc-dkPtyc cVhNnt"
        >
          <img
            alt="[든든] 유부 슬라이스 500g"
            class="sc-hKwCoD doJTuW"
            src="https://cdn-mart.baemin.com/goods/custom/20200525/11153-main-01.png"
          />
        </div>
      </div>
      <div
        class="sc-jRQAMF liUEKP"
      >
        <h4
          class="sc-gKckTs jFQqTJ"
        >
          [든든] 유부 슬라이스 500g
        </h4>
        <div
          class="sc-iCfLBT bWXkrK"
        >
          <div
            class="sc-ieebsP jwDezp"
          >
            <span
              class="sc-dJjZJu ezWYrJ"
            >
              4,900
               원
            </span>
            <button
              aria-label="[든든] 유부 슬라이스 500g을 장바구니에 담기"
              class="sc-bkkfTU jprmXP"
              type="submit"
            >
              <svg>
                cart.svg
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-eCImvq kFPXzk"
    >
      <div
        class="sc-gsDJrp"
      >
        <div
          class="sc-dkPtyc cVhNnt"
        >
          <img
            alt="[든든] 진맛살 1kg"
            class="sc-hKwCoD doJTuW"
            src="https://cdn-mart.baemin.com/goods/custom/20200525/11157-main-01.png"
          />
        </div>
      </div>
      <div
        class="sc-jRQAMF liUEKP"
      >
        <h4
          class="sc-gKckTs jFQqTJ"
        >
          [든든] 진맛살 1kg
        </h4>
        <div
          class="sc-iCfLBT bWXkrK"
        >
          <div
            class="sc-ieebsP jwDezp"
          >
            <span
              class="sc-dJjZJu ezWYrJ"
            >
              7,100
               원
            </span>
            <button
              aria-label="[든든] 진맛살 1kg을 장바구니에 담기"
              class="sc-bkkfTU jprmXP"
              type="submit"
            >
              <svg>
                cart.svg
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-eCImvq kFPXzk"
    >
      <div
        class="sc-gsDJrp"
      >
        <div
          class="sc-dkPtyc cVhNnt"
        >
          <img
            alt="[든든] 신-키리모찌 1kg"
            class="sc-hKwCoD doJTuW"
            src="https://cdn-mart.baemin.com/goods/custom/20200525/11161-main-01.png"
          />
        </div>
      </div>
      <div
        class="sc-jRQAMF liUEKP"
      >
        <h4
          class="sc-gKckTs jFQqTJ"
        >
          [든든] 신-키리모찌 1kg
        </h4>
        <div
          class="sc-iCfLBT bWXkrK"
        >
          <div
            class="sc-ieebsP jwDezp"
          >
            <span
              class="sc-dJjZJu ezWYrJ"
            >
              12,000
               원
            </span>
            <button
              aria-label="[든든] 신-키리모찌 1kg을 장바구니에 담기"
              class="sc-bkkfTU jprmXP"
              type="submit"
            >
              <svg>
                cart.svg
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-eCImvq kFPXzk"
    >
      <div
        class="sc-gsDJrp"
      >
        <div
          class="sc-dkPtyc cVhNnt"
        >
          <img
            alt="[든든] 흑곤약 250g"
            class="sc-hKwCoD doJTuW"
            src="https://cdn-mart.baemin.com/goods/custom/20200525/11263-main-01.png"
          />
        </div>
      </div>
      <div
        class="sc-jRQAMF liUEKP"
      >
        <h4
          class="sc-gKckTs jFQqTJ"
        >
          [든든] 흑곤약 250g
        </h4>
        <div
          class="sc-iCfLBT bWXkrK"
        >
          <div
            class="sc-ieebsP jwDezp"
          >
            <span
              class="sc-dJjZJu ezWYrJ"
            >
              1,300
               원
            </span>
            <button
              aria-label="[든든] 흑곤약 250g을 장바구니에 담기"
              class="sc-bkkfTU jprmXP"
              type="submit"
            >
              <svg>
                cart.svg
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-eCImvq kFPXzk"
    >
      <div
        class="sc-gsDJrp"
      >
        <div
          class="sc-dkPtyc cVhNnt"
        >
          <img
            alt="[든든] 흑곤약 1.2kg"
            class="sc-hKwCoD doJTuW"
            src="https://cdn-mart.baemin.com/goods/custom/20200525/11277-main-01.png"
          />
        </div>
      </div>
      <div
        class="sc-jRQAMF liUEKP"
      >
        <h4
          class="sc-gKckTs jFQqTJ"
        >
          [든든] 흑곤약 1.2kg
        </h4>
        <div
          class="sc-iCfLBT bWXkrK"
        >
          <div
            class="sc-ieebsP jwDezp"
          >
            <span
              class="sc-dJjZJu ezWYrJ"
            >
              4,000
               원
            </span>
            <button
              aria-label="[든든] 흑곤약 1.2kg을 장바구니에 담기"
              class="sc-bkkfTU jprmXP"
              type="submit"
            >
              <svg>
                cart.svg
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-eCImvq kFPXzk"
    >
      <div
        class="sc-gsDJrp"
      >
        <div
          class="sc-dkPtyc cVhNnt"
        >
          <img
            alt="[든든] 곤약말이 시라타끼 200g"
            class="sc-hKwCoD doJTuW"
            src="https://cdn-mart.baemin.com/goods/custom/20200525/11280-main-01.png"
          />
        </div>
      </div>
      <div
        class="sc-jRQAMF liUEKP"
      >
        <h4
          class="sc-gKckTs jFQqTJ"
        >
          [든든] 곤약말이 시라타끼 200g
        </h4>
        <div
          class="sc-iCfLBT bWXkrK"
        >
          <div
            class="sc-ieebsP jwDezp"
          >
            <span
              class="sc-dJjZJu ezWYrJ"
            >
              1,100
               원
            </span>
            <button
              aria-label="[든든] 곤약말이 시라타끼 200g을 장바구니에 담기"
              class="sc-bkkfTU jprmXP"
              type="submit"
            >
              <svg>
                cart.svg
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-eCImvq kFPXzk"
    >
      <div
        class="sc-gsDJrp"
      >
        <div
          class="sc-dkPtyc cVhNnt"
        >
          <img
            alt="[든든] 기분 시로가마보코 160g"
            class="sc-hKwCoD doJTuW"
            src="https://cdn-mart.baemin.com/goods/custom/20200525/11315-main-01.png"
          />
        </div>
      </div>
      <div
        class="sc-jRQAMF liUEKP"
      >
        <h4
          class="sc-gKckTs jFQqTJ"
        >
          [든든] 기분 시로가마보코 160g
        </h4>
        <div
          class="sc-iCfLBT bWXkrK"
        >
          <div
            class="sc-ieebsP jwDezp"
          >
            <span
              class="sc-dJjZJu ezWYrJ"
            >
              2,800
               원
            </span>
            <button
              aria-label="[든든] 기분 시로가마보코 160g을 장바구니에 담기"
              class="sc-bkkfTU jprmXP"
              type="submit"
            >
              <svg>
                cart.svg
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-eCImvq kFPXzk"
    >
      <div
        class="sc-gsDJrp"
      >
        <div
          class="sc-dkPtyc cVhNnt"
        >
          <img
            alt="[든든] 냉동조미유부삼각 (60장) 1kg"
            class="sc-hKwCoD doJTuW"
            src="https://cdn-mart.baemin.com/goods/custom/20200525/11323-main-01.png"
          />
        </div>
      </div>
      <div
        class="sc-jRQAMF liUEKP"
      >
        <h4
          class="sc-gKckTs jFQqTJ"
        >
          [든든] 냉동조미유부삼각 (60장) 1kg
        </h4>
        <div
          class="sc-iCfLBT bWXkrK"
        >
          <div
            class="sc-ieebsP jwDezp"
          >
            <span
              class="sc-dJjZJu ezWYrJ"
            >
              8,400
               원
            </span>
            <button
              aria-label="[든든] 냉동조미유부삼각 (60장) 1kg을 장바구니에 담기"
              class="sc-bkkfTU jprmXP"
              type="submit"
            >
              <svg>
                cart.svg
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-eCImvq kFPXzk"
    >
      <div
        class="sc-gsDJrp"
      >
        <div
          class="sc-dkPtyc cVhNnt"
        >
          <img
            alt="[든든] 시노다마끼 유부야채말이 (15g*40입) 600g"
            class="sc-hKwCoD doJTuW"
            src="https://cdn-mart.baemin.com/goods/custom/20200525/11328-main-01.png"
          />
        </div>
      </div>
      <div
        class="sc-jRQAMF liUEKP"
      >
        <h4
          class="sc-gKckTs jFQqTJ"
        >
          [든든] 시노다마끼 유부야채말이 (15g*40입) 600g
        </h4>
        <div
          class="sc-iCfLBT bWXkrK"
        >
          <div
            class="sc-ieebsP jwDezp"
          >
            <span
              class="sc-dJjZJu ezWYrJ"
            >
              9,000
               원
            </span>
            <button
              aria-label="[든든] 시노다마끼 유부야채말이 (15g*40입) 600g을 장바구니에 담기"
              class="sc-bkkfTU jprmXP"
              type="submit"
            >
              <svg>
                cart.svg
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-eCImvq kFPXzk"
    >
      <div
        class="sc-gsDJrp"
      >
        <div
          class="sc-dkPtyc cVhNnt"
        >
          <img
            alt="[든든] 기분 오뎅세트 433g"
            class="sc-hKwCoD doJTuW"
            src="https://cdn-mart.baemin.com/goods/custom/20200525/11331-main-01.png"
          />
        </div>
      </div>
      <div
        class="sc-jRQAMF liUEKP"
      >
        <h4
          class="sc-gKckTs jFQqTJ"
        >
          [든든] 기분 오뎅세트 433g
        </h4>
        <div
          class="sc-iCfLBT bWXkrK"
        >
          <div
            class="sc-ieebsP jwDezp"
          >
            <span
              class="sc-dJjZJu ezWYrJ"
            >
              8,900
               원
            </span>
            <button
              aria-label="[든든] 기분 오뎅세트 433g을 장바구니에 담기"
              class="sc-bkkfTU jprmXP"
              type="submit"
            >
              <svg>
                cart.svg
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
